src/python/tests/test_maf_to_fasta.py

```diff
@@ -0,0 +1,131 @@
+            (['genomeA', 'genomeA.1'], None, raises(ValueError)),
+            ([], None, raises(ValueError))
+        ]
+    )
+    def test_compile_maf_src_regex(self, genome_names: Iterable[str], exp_output: Pattern[str],
+                                   expectation: ContextManager) -> None:
+        """Tests :func:`maf_to_fasta.compile_maf_src_regex()` function.
+
+        Args:
+            genome_names: The genome names expected to be in the input MAF file.
+            exp_output: Expected return value of the function.
+            expectation: Context manager for the expected exception. The test will only pass if that
+                exception is raised. Use :class:`~contextlib.nullcontext` if no exception is expected.
+
+        """
+        with expectation:
+            obs_output = maf_to_fasta.compile_maf_src_regex(genome_names)
+            assert obs_output == exp_output
+
+    @pytest.mark.parametrize(
+        "maf_file, output_dir, genomes_file, out_file_rel_paths, expectation",
+        [
+            ('gabs.maf', 'gabs', 'gab_genomes.txt', ['0.fa', '0.json', '1.fa', '1.json'], does_not_raise()),
+            ('gabs.maf', 'gabs', 'wrong_genomes.txt', [], raises(ValueError)),
+            ('gabs.maf', 'gabs', None, [], raises(ValueError))
+        ]
+    )
+    def test_convert_maf_to_fasta(self, maf_file: PathLike, output_dir: PathLike,
+                                  genomes_file: Optional[PathLike], out_file_rel_paths: List[str],
+                                  expectation: ContextManager, tmp_dir: Path) -> None:
+        """Tests :func:`maf_to_fasta.convert_maf_to_fasta()` function.
+
+        Args:
+            maf_file: Input MAF file with alignment blocks. The src fields of
+                this MAF file should be of the form '<genome>.<seqid>'.
+            output_dir: Output directory under which FASTA and JSON files will be created.
+            genomes_file: File listing the genomes in the input MAF file, one per line.
+            out_file_rel_paths: Relative paths of the output FASTA and JSON files.
+            expectation: Context manager for the expected exception. The test will only pass if that
+                exception is raised. Use :class:`~contextlib.nullcontext` if no exception is expected.
+            tmp_dir: Unit test temp directory (fixture).
+
+        """
+        maf_file_path = self.ref_file_dir / maf_file
+        out_dir_path = tmp_dir / output_dir
+        if genomes_file is not None:
+            genomes_file = self.ref_file_dir / genomes_file
+
+        with expectation:
+            maf_to_fasta.convert_maf_to_fasta(maf_file_path, out_dir_path, genomes_file=genomes_file)
+
+        ref_dir_path = self.ref_file_dir / output_dir
+        for out_file_rel_path in out_file_rel_paths:
+            out_file_path = out_dir_path / out_file_rel_path
+            ref_file_path = ref_dir_path / out_file_rel_path
+            assert filecmp.cmp(out_file_path, ref_file_path)
```

Comment:
The `assert` bit still needs to go inside the `with`, I would say, to ensure it doesn't fail (not sure what the result of the test would be if the `with` passes but the `assert` fails).

Reply:
Currently, the `with` passes only for the first test case. If the `assert` fails then an `AssertionError` is raised, but this will only happen if the files do not match, indicating that the observed output doesn't match the expected output. I suppose we could set `expectation` to `raises(AssertionError)` if we had a test case where we expected a mismatch between the observed/expected output. But in that case, wouldn't the expected output be... unexpected? 🤨